### PR TITLE
Break HTML blocks to individual strings

### PR DESCRIPTION
**Pull request by kaakaa**
#### Summary
(Some messages have not been fixed yet. See the bottom of Issue for details.)
Broke messages containing HTML tag to individual messages, and modifyed accordingly other resources such as e-mail templates etc.

#### Ticket Link
#6174
[JIRA](https://mattermost.atlassian.net/browse/MM-3736)

#### Checklist
- [x] Includes text changes and localization file ([.../i18n/en.json](https://github.com/mattermost/mattermost-server/blob/master/i18n/en.json)) updates

---

## Status

UPDATED: 2018/06/27

| No | Status  | Message ID |
|:--:|:--------|:-----------|
| 1  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/298194723a8d8555bcb304e1a147d85f888a6c4f)       | api.admin.test_email.body |
| 2  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/6a014c90a304521a71f5e6f30ff0db07947dae79)    | api.templates.email_change_body.info |
| 3  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/4b510c95d1ce754b5e418a2b8594ba30dca06fcd)    | api.templates.email_info |
| 4  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/0038ed058cee4626a35b327f25269fa3bd51ae63), ([FIX](https://github.com/mattermost/mattermost-server/pull/8903/commits/15114b48df8d18a8304a60bcd6586f1d177bb611))   | api.templates.deactivate_body.info |
| 5  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/36c9ca7b881310abf2be8290b855bdaa1cafe9dc) | api.templates.invite_body.extra_info |
| 6  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/36c9ca7b881310abf2be8290b855bdaa1cafe9dc) | api.templates.invite_body.info |
| 7  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/50e038c1f6a5782f3465b1b1ece47e1a7f6b7031)    | api.templates.mfa_activated_body.info |
| 8  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/50e038c1f6a5782f3465b1b1ece47e1a7f6b7031)      | api.templates.mfa_deactivated_body.info |
| 9  | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/a7109f63edf8f61d1cbea9491d4560a75f4de0f9)    | api.templates.password_change_body.info |
| 10 | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/579b6fb62bdc2705f84105940c83a8ab97d3d4e1)    | api.templates.reset_body.info |
| 11 | [DONE](hhttps://github.com/mattermost/mattermost-server/pull/8903/commits/67f9236750d6f321d54ef6de29a835f2828e47e4)    | api.templates.signin_change_email.body.info |
| 12 | Removed by #8980 | api.templates.signup_team_body.info |
| 13 | Removed by #8980  | api.templates.upgrade_30_body.info |
| 14 | DONE [1](https://github.com/mattermost/mattermost-server/pull/8903/commits/c64b7be936174eb51e3332820d915b1cf0fe61fc),[2](https://github.com/mattermost/mattermost-server/pull/8903/commits/b58b31b05ba9c80268b60683154e9b2d020d7ca9),[3](https://github.com/mattermost/mattermost-server/pull/8903/commits/fd8be08aade2df60ba350e293379a38f3c904118)    | api.templates.user_access_token_body.info |
| 15 | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/9ce55d3c62b120ad8693af095f00990088e9325b)    | api.templates.username_change_body.info |
| 16 | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/5fe56efc8310a23b378780bac34440f2413bb2a1)    | app.notification.body.text.group_message.full |
| 17 | [DONE](https://github.com/mattermost/mattermost-server/pull/8903/commits/5fe56efc8310a23b378780bac34440f2413bb2a1)    | app.notification.body.text.notification.full |

## Questions
* ~~No.1~~
  * ~~This message is used in test email directly. Can I just remove 3 `<br/>` tags from the message?~~
* ~~No.5,6~~
  * ~~HTML tags appear in the middle of the message, so it seems necessary to change the whole message. But I have no idea how should we change these message.~~

## NOTE
* No.2, 16
  * This message has been fixed already. But the last sentence (`If you did not make this change,~`) is changed to `If this change wasn't initiated by you, please contact your system administrator.` to make it the same as other messages.
* No.14: The following changes include.
  * Add Props["siteURL"] (to fix bug)
  * Add argument `siteURL` to make it same as other methods in email.go

### CODE REVIEW

File: i18n/en.json [hunk 1]
```diff
@@ -2664,27 +2680,31 @@
...
     "id": "api.templates.mfa_activated_body.info",
-    "translation": "Multi-factor authentication has been added to your account on {{ .SiteURL }}.<br>If this change wasn't initiated by you, please contact your system administrator."
+    "translation": "Multi-factor authentication has been added to your account on {{ .SiteURL }}."
   },
   {
     "id": "api.templates.mfa_activated_body.title",
     "translation": "Multi-factor authentication was added"
   },
+  {
+    "id": "api.templates.mfa_change_extra",
```
Comment (jasonblais) [inline i18n/en.json]: Wondering if we can combine this with `api.templates.mfa_change_extra` so we don't have a duplicate string to translate?

Comment (kaakaa) [reply i18n/en.json]: Sorry, this is my mistakes.
`api.templates.mfa_change_extra` has never been used anywhere, so it should be removed.

**Comment by kaakaa** [comment]
> https://github.com/kaakaa/mattermost-server/blob/8a8b3959fe3136fb03cd5e30f64361b856401674/i18n/en.json#L144

[Fixed this one](https://github.com/mattermost/mattermost-server/pull/8903/commits/176cb118d3822f1bd33c97ea644b918c1b5fca1d).

> https://github.com/kaakaa/mattermost-server/blob/8a8b3959fe3136fb03cd5e30f64361b856401674/i18n/en.json#L2667
> https://github.com/kaakaa/mattermost-server/blob/8a8b3959fe3136fb03cd5e30f64361b856401674/i18n/en.json#L2671

HTML tags(<strong>) appear in the middle of these message, so it seems necessary to change the whole message. But I have no ideas how should we change these message.
Do you have any ideas?

---

> Are you considering of doing the same for the webapp and mobile repositories?

Yes, I'm thinking about doing the same to webapp/mobile.
But in webapp, there are over 100 messages with html tag. So It may also take a little while.

**Comment by jasonblais** [comment]
> HTML tags() appear in the middle of these message, so it seems necessary to change the whole message. But I have no ideas how should we change these message.
Do you have any ideas?

You mean the `<strong>` tags? That's true. @enahum Any thoughts on how we might be able to remove/eliminate them from the localized messages?

> Yes, I'm thinking about doing the same to webapp/mobile.
But in webapp, there are over 100 messages with html tag. So It may also take a little while.

Yeah, absolutely understand. Thanks so much for your help on this 👍

**Comment by kaakaa** [comment]
@jasonblais Yes, I mean `<strong>` tag. I fixed that comment, thanks.

**Comment by jasonblais** [comment]
> HTML tags() appear in the middle of these message, so it seems necessary to change the whole message. But I have no ideas how should we change these message.
Do you have any ideas?

You mean the `<strong>` tags? That's true. @enahum Any thoughts on how we might be able to remove/eliminate them from the localized messages?

**Comment by jasonblais** [comment]
@enahum ^

**Comment by enahum** [comment]
The changes looks good to me, now for 5 and 6 those are tricky ones maybe we can do when building the template?

something like:
the string to translate is 
`The team {{.SenderStatus}} [[.SenderName]], has invited you to join [[.TeamDisplayName]]`

and then in code we replace the `[[` for `<strong>{{` and `]]` for `}}</strong>` 

and we treat that as a new convention.

how that sounds?

**Comment by kaakaa** [comment]
Thanks @enahum. it sounds good.
If new convension is going, I'll change `TranslateAsHtml` to `TranslateAsMarkup` and add logic to replace `[[` and `]]` to `<strong>{{` and `}}</strong>` respectively.

**Comment by kaakaa** [comment]
Hi @enahum, @jasonblais
I looked into `i18n.TranslateFunc` and found that I cannot insert the process to replace `[[`, `]]` in messages with `<strong>{{` and `}}</strong>` respectively.
It seems that translation functions cannot receive **message string** to be translated, it can receive only translation **ID**.

Do you know code that is doing something similar to what we want to do?

---

Otherwise, we can also divide that messages by `<strong>` tags and move `<strong>{{.XXX}}</strong>` to templates.

Divide **No.5**: `Mattermost lets you share messages and files from your PC or phone, with instant search and archiving. After you’ve joined <strong>{{.TeamDisplayName}}</strong>, you can sign-in to your new team and access these features anytime from the web address:<br/><br/><a href='{{.TeamURL}}'>{{.TeamURL}}</a>`  into the following three messages.
  * "Mattermost lets you share messages and files from your PC or phone, with instant search and archiving."
  * "After you’ve joined in"
  * "you can sign-in to your new team and access these features anytime from the web address:"

Divide **No.6**: `The team {{.SenderStatus}} <strong>{{.SenderName}}</strong>, has invited you to join <strong>{{.TeamDisplayName}}</strong>.` into the following two messages
  * "The team {{.SenderStatus}}"
  * "has invited you in "

It may be difficult for some translator to translate these messages.
how's this sounds?

**Comment by enahum** [comment]
So my thought process was like this:
Key: somekey
Translation: a string in the translated language that includes brackets.

The you get the translated string by calling the the T function and after that you replace the double brackets with the html tag, no need to modify the translated function

**Comment by kaakaa** [comment]
oh...you are right, I have misunderstood about translation function. Thank you!
I'll implement and push it later.

**Comment by kaakaa** [comment]
hmm...I'm sorry, I get confused.

I think that T function embed parameter value in `{{.SenderName}}`.
So even if we replace `[[.XXX]]` with `<strong>{{.XXX}}</strong>` after getting translated string, we cannot embed parameter value in `{{.XXX}}`.

How do I embed the value in `{{.XXX}}`?

**Comment by enahum** [comment]
Ok so lets look at it this way:

string to be translated: key: "awesome.channel",  translation: "Hey men check out [[{{.ChannelName}}]] is awesome"
then you would call the T function like `result := utils.T("awesome.channel")`

then result is equals to the translation with the var replaced like: **"Hey man check out [[My new channel]] is awesome"**

and you take that result in bold and replace the brackets with the `<strong>` tags.

make sense?

**Comment by kaakaa** [comment]
@jasonblais @enahum 
I've done all things about this issue.
Please review this.

**Comment by jasonblais** [comment]
Thanks @kaakaa! Our QA team will do a brief test of email notifications and then we should be good for code review.

**Comment by jasonblais** [comment]
@lindalumitchell @lindy65 Quick reminder to help with this PR when your team has a chance :)

**Comment by lindy65** [comment]
@jasonblais @kaakaa 

Thanks, I've tested email notifications on the spinmint, specifically
- for sending "immediately" or "never" 
- with batching on and off
- for at-mentions, direct messages, group messages 
- for all variations of "words that trigger mentions"

I did not find any issues and email notifications worked as expected in all cases.

**Comment by jasonblais** [comment]
Thanks Lindy! Moving to dev review next.

File: app/email.go [hunk 2]
```diff
@@ -26,8 +26,9 @@ func (a *App) SendChangeUsernameEmail(oldUsername, newUsername, email, locale, s
 	bodyPage := a.NewEmailTemplate("email_change_body", locale)
 	bodyPage.Props["SiteURL"] = siteURL
 	bodyPage.Props["Title"] = T("api.templates.username_change_body.title")
-	bodyPage.Html["Info"] = utils.TranslateAsHtml(T, "api.templates.username_change_body.info",
+	bodyPage.Props["Info"] = T("api.templates.username_change_body.info",
 		map[string]interface{}{"TeamDisplayName": a.Config().TeamSettings.SiteName, "NewUsername": newUsername})
+	bodyPage.Props["Warning"] = T("api.templates.email_warning")
```
Comment (jespino) [inline app/email.go]: `api.templates.email_warning` is not defined in the translation keys (I think)

Comment (kaakaa) [reply app/email.go]: Yes, you are right. it's missing in `i18n/en.json`.
`api.templates.email_warning` should be `If you did not make this change, please contact the system administrator.`. I'll add it.

Thank you.

**Comment by enahum** [comment]
@kaakaa there are some build failures and the PR needs rebase, once that's done I think it is safe to merge it

**Comment by jasonblais** [comment]
@enahum @jespino this is ready to merge

**Comment by enahum** [comment]
@jasonblais merged

**Pull request merged**